### PR TITLE
Backend CSV generation Python

### PR DESCRIPTION
**Pull request by chenrichard10**
## Notion ticket link
Create backend utility to generate csv from tabular data, add option to return csv from bulk GET endpoints
[Ticket Name](https://www.notion.so/uwblueprintexecs/Create-backend-utility-to-generate-csv-from-tabular-data-add-option-to-return-csv-from-bulk-GET-end-19dd604265034b5b8cfe45a9f1aea84d)

## Implementation description
*  CSV generation for Python with unwind and flattening functionality. Pytests tests have been written as well.

## Steps to test
1. REST (using Postman)

    Set the content-type in the headers to text/csv and the request type to GET on entities and users.

## What should reviewers focus on?
*  Unsure if the csv string is the exact output we want
*  CSV formatting functionality in generate_csv.py

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/python/app/utilities/generate_csv.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+Some Notes: 
+1. Unwind only unwinds a single level (i.e a list)
+2. CSV requires all dictionaries in the list are of the same type
+"""
+
+import csv
+import io
+
+
+def flatten_objects(dictionary, parent_key="", sep="."):
```
Comment (sherryhli) [inline backend/python/app/utilities/generate_csv.py:12]: Nit: could we name this `flatten_dict` instead?

File: backend/python/app/utilities/generate_csv.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    Flatten a dictionary of dictionaries into a single dictionary
+    :param dictionary: dictionary to flatten
+    :param parent_key: key to use as the parent key for the new dictionary
+    :param sep: separator to use when combining keys
+    :return: flattened dictionary
+    :rtype: dict
+    """
+    items = []
+    for key, val in dictionary.items():
+        new_key = parent_key + sep + key if parent_key else key
```
Comment (sherryhli) [inline backend/python/app/utilities/generate_csv.py:23]: Should we account for non-string keys? Maybe cast `key` to str

File: backend/python/app/utilities/generate_csv.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        if isinstance(val, dict):
+            items.extend(flatten_objects(val, new_key, sep=sep).items())
+
+        else:
+            items.append((new_key, val))
+    return dict(items)
+
+
+def flatten_lists(dictionary, sep="."):
```
Comment (sherryhli) [inline backend/python/app/utilities/generate_csv.py:32]:
```suggestion
def flatten_lists_in_dict(dictionary, sep="."):
```

File: backend/python/app/utilities/generate_csv.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+
+def unwind_field(list_of_dict, field):
+    """
+    Unwinds lists inside dicts into multiple dictionaries, returning a new list at the end
+    :param list_of_dict: list of dictionaries
+    :param field: field to unwind
+    :return: list of dictionaries
+    :rtype: list of dict
+    """
```
Comment (sherryhli) [inline backend/python/app/utilities/generate_csv.py:73]: I know you have tests but would be nice to add an example in the docstrings, took me a while to understand what unwind means

File: backend/typescript/graphql/resolvers/entityResolvers.ts [hunk 5]
```diff
@@ -8,6 +8,7 @@ import {
   validateFileType,
   getFileTypeValidationError,
 } from "../../middlewares/validators/util";
+import { generateCSV } from "../../utilities/CSVUtils";
```
Comment (sherryhli) [inline backend/typescript/graphql/resolvers/entityResolvers.ts:11]: Can you please revert all the TS changes? Will merge the TS implementation with #91.

File: docker-compose.yml [hunk 6]
```diff
@@ -1,35 +1,35 @@
...
-    ports:
-      - 5000:5000
-    dns:
-      - 8.8.8.8
-    depends_on:
-      - db
-      - db-test
-    env_file:
-      - ./.env
+  # frontend:
```
Comment (sherryhli) [inline docker-compose.yml:4]: Uncomment please

File: backend/python/app/rest/entity_routes.py [hunk 7]
```diff
@@ -44,11 +60,16 @@ def get_entities():
...
+        content_type = request.mimetype
     except Exception as e:
         error_message = getattr(e, "message", None)
         return jsonify({"error": (error_message if error_message else str(e))}), 500
 
     # HTTP status code 200 means OK
+    if content_type == "text/csv":
+        result = entity_service.get_entities()
+        return jsonify(generate_csv_from_list(result)), 200
+
```
Comment (sherryhli) [inline backend/python/app/rest/entity_routes.py:72]: Sorry for the misunderstanding but we don't need CSV functionality for single entities, just the bulk GET request above is enough :)

**Review by sherryhli — CHANGES_REQUESTED** [review]
Good progress, thanks for creating all the utility functions from scratch and writing tests!

File: backend/python/app/utilities/generate_csv.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
+"""
```
Comment (sherryhli) [inline backend/python/app/utilities/generate_csv.py:1]: Let's rename this file `csv_utils.py` for consistency with the TS backend

File: backend/python/app/utilities/test_csv.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
+"""
```
Comment (sherryhli) [inline backend/python/app/utilities/test_csv.py:1]: Can we move this test file to be under `backend/python/tests` instead?

**Review by sherryhli — COMMENTED** [review]
Two other nits

File: backend/python/app/utilities/test_csv.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+
+from generate_csv import generate_csv_from_list
+
+
+def test_basic():
+    result = generate_csv_from_list(person)
+    print(result)
+    assert (
+        result
+        == "Person1,20,\"[{'name': 'Beans', 'type': 'Cat'}, {'name': 'Spot', 'type': 'Dog'}]\"\r\nPerson2,25,\"[{'name': 'Splash', 'type': 'Fish'}]\"\r\n"
```
[the remark this answers is not in the document]
Comment (chenrichard10) [reply backend/python/app/utilities/test_csv.py:113]: \n does not work by itself

File: backend/python/app/utilities/test_csv.py (hunk 9 above)
Comment (chenrichard10) [reply backend/python/app/utilities/test_csv.py:1]: Moved to functional folder

Comment (sherryhli) [reply backend/python/app/utilities/test_csv.py]: Hmm, I think `unit` might be more suitable

**Review by sherryhli — DISMISSED** [review]
LGTM! I left a couple of nit comments, no need to address them. I'll likely make those small changes when I merge this into `main`.

**Review by sherryhli — APPROVED** [review]

**Pull request merged**